include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      set_sensitivity_value(const scalar_type &value, const unsigned int index);
+
+      /**
+       * Initialise an independent variable $X_{i}$ such that subsequent
+       * operations performed with it are tracked.
+       *
+       * @note Care must be taken to mark each independent variable only once.
+       *
+       * @note The order in which the independent variables are marked defines the
+       * order of all future internal operations. They must be manipulated in
+       * the same order as that in which they are first marked. If not then,
+       * for example, ADOL-C won't throw an error, but rather it might complain
+       * nonsensically during later computations or produce garbage results.
+       *
+       * @param[out] out An auto-differentiable number that is ready for use in
+       * computations. The operations that are performed with it are recorded on
+       * the tape and will be considered in the computation of dependent
+       * variable values.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      mark_independent_variable(ad_type &out, const unsigned int index) const;
+
+      /**
+       * Finalize the state of the independent variables before use.
+       *
+       * This step and the storage of the independent variables is done
+       * separately because some derived classes may offer the capability
+       * to add independent variables in a staggered manner. This function
+       * is to be triggered when these values are considered finalized
+       * and we can safely initialize the sensitive equivalents of those
+       * values.
+       */
+      void
+      finalize_sensitive_independent_variables() const;
+
+      /**
+       * Initialize an independent variable $X_{i}$.
+       *
+       * @param[out] out An auto-differentiable number that is ready for use in
+       * standard computations. The operations that are performed with it are
+       * not recorded on the tape, and so should only be used when not in
+       * recording mode.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      initialize_non_sensitive_independent_variable(
+        ad_type &          out,
+        const unsigned int index) const;
+
+      /**
+       * The number of independent variables that have been manipulated within a
+       * set of operations.
+       */
+      unsigned int
+      n_touched_independent_variables() const;
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * The set of dependent variables $\mathbf{f}(\mathbf{X})$ of which the
+       * derivatives with respect to $\mathbf{X}$ will be computed.
+       *
+       * The gradients and Hessians of these dependent variables will be
```

Comment:
Somewhere higher up in this file, I saw you spell "hessian" in lower-case letters. I prefer "Hessian" and "Jacobian". In the end, that's not important, but while you're editing around here, choose a common spelling.